File: src/devilry_subjectadmin/devilry_subjectadmin/static/devilry_subjectadmin/app/model/Assignment.js
/** Assignment model. */
Ext.define('devilry_subjectadmin.model.Assignment', {
    extend: 'Ext.data.Model',

    requires: [
        'Ext.Date',
        'devilry_extjsextras.DatetimeHelpers'
    ],

    idProperty: 'id',
    fields: [
        {name: 'admins',  type: 'auto'},
        {name: 'breadcrumb',  type: 'auto'},
        {name: 'can_delete',  type: 'bool'},
        {name: 'anonymous',  type: 'bool'},
        {name: 'etag',  type: 'string'},
        {name: 'first_deadline',  type: 'date', "dateFormat": "Y-m-d H:i:s"},
        {name: 'id', type: 'auto'},
        {name: 'inherited_admins',  type: 'auto'},
        {name: 'long_name',  type: 'string'},
        {name: 'parentnode', type: 'auto'},
        {name: 'publishing_time',  type: 'date', "dateFormat": "Y-m-d H:i:s"},
        {name: 'is_published',  type: 'bool', persist: false},
        {name: 'publishing_time_offset_from_now',  type: 'auto', persist: false},
        {name: 'short_name',  type: 'string'},
        {name: 'scale_points_percent',  type: 'int'},
        {name: 'deadline_handling',  type: 'int'},
        {name: 'delivery_types',  type: 'int'},
        {name: 'number_of_groups',  type: 'int', persist: false},
        {name: 'number_of_deliveries',  type: 'int', persist: false},
        {name: 'number_of_candidates',  type: 'int', persist: false},
        {name: 'number_of_groups_where_is_examiner',  type: 'int', persist: false},
        {name: 'has_valid_grading_setup', type: 'bool', persist: false},
        {name: 'gradingsystemplugin_title', type: 'string', persist: false},
        {name: 'max_points', type: 'int'},
        {name: 'passing_grade_min_points', type: 'int'},
        {name: 'points_to_grade_mapper', type: 'string'},
        {name: 'grading_system_plugin_id', type: 'string'}
    ],

    proxy: {
        type: 'rest',
        url: DevilrySettings.DEVILRY_URLPATH_PREFIX + '/devilry_subjectadmin/rest/assignment/',
        extraParams: {
            format: 'json'
        },
        reader: {
            type: 'json'
        }
    },

    formatPublishOffsetFromNow: function() {
        return devilry_extjsextras.DatetimeHelpers.formatTimedeltaRelative(
            this.get('publishing_time_offset_from_now'), !this.get('is_published'));
    },

    formatPublishingTime: function() {
        return devilry_extjsextras.DatetimeHelpers.formatDateTimeShort(this.get('publishing_time'));
    },
    formatFirstDeadline: function() {
        return devilry_extjsextras.DatetimeHelpers.formatDateTimeShort(this.get('first_deadline'));
    },


    /** Get period info from breadcrumb.
     *
     * @return {Object} An object with ``id``, ``path`` and ``is_admin``. If ``is_admin`` is ``false``, ``id`` is ``null``.
     *
     * */
    getPeriodInfoFromBreadcrumb: function() {
        var breadcrumb = this.get('breadcrumb');
        var period_id;
        var periodpath;
        var is_admin;
        if(breadcrumb.length === 0) {
         throw "Breadcrumb is empty";
        } else if(breadcrumb.length === 1) {
            period_id = null;
            periodpath = breadcrumb[0].text.split('.').slice(0, 2).join('.');
            is_admin = false;
        } else {
            var periodpathArray = [];
            for(var index=0; index<breadcrumb.length-1; index++)  { // Exclude the last item (the assignment)
                var breadcrumbItem = breadcrumb[index];
                periodpathArray.push(breadcrumbItem.text);
            }
            period_id = breadcrumb[breadcrumb.length - 2].id;
            periodpath = periodpathArray.join('.');
            is_admin = true;
        }
        return {
            id: period_id,
            path: periodpath,
            is_admin: is_admin
        };
    }
});

File: devilry/devilry_subjectadmin/rest/assignment.py
from datetime import datetime

from django.db.models import Count

from djangorestframework.permissions import IsAuthenticated
from devilry.apps.core.models import Assignment
from devilry.apps.core.models import Examiner
from devilry.devilry_rest.serializehelpers import format_datetime
from devilry.devilry_rest.serializehelpers import format_timedelta
from devilry.devilry_gradingsystem.pluginregistry import GradingSystemPluginNotInRegistryError
from .auth import IsAssignmentAdmin
from .auth import periodadmin_required
from devilry.devilry_subjectadmin.rest.viewbase import BaseNodeInstanceModelView
from devilry.devilry_subjectadmin.rest.viewbase import BaseNodeListOrCreateView
from .resources import BaseNodeInstanceResource


class AssignmentResourceMixin(object):
    def publishing_time(self, instance):
        if isinstance(instance, self.model):
            return format_datetime(instance.publishing_time)

    def publishing_time_offset_from_now(self, instance):
        if isinstance(instance, self.model):
            return format_timedelta(datetime.now() - instance.publishing_time)

    def is_published(self, instance):
        if isinstance(instance, self.model):
            return instance.publishing_time < datetime.now()

    def first_deadline(self, instance):
        if isinstance(instance, self.model) and instance.first_deadline:
            return format_datetime(instance.first_deadline)



class AssignmentResource(AssignmentResourceMixin, BaseNodeInstanceResource):
    model = Assignment
    fields = ('id', 'parentnode', 'short_name', 'long_name', 'etag',
              'publishing_time', 'delivery_types', 'scale_points_percent',
              'is_published', 'publishing_time_offset_from_now',
              'first_deadline', 'anonymous', 'deadline_handling')

class AssignmentInstanceResource(AssignmentResourceMixin, BaseNodeInstanceResource):
    model = Assignment
    fields = AssignmentResource.fields + (
        'can_delete', 'admins', 'inherited_admins', 'breadcrumb', 'number_of_groups',
        'number_of_deliveries', 'number_of_groups_where_is_examiner', 'number_of_candidates',
        'gradingsystemplugin_title', 'has_valid_grading_setup',
        'max_points', 'passing_grade_min_points', 'points_to_grade_mapper',
        'grading_system_plugin_id')

    def _serialize_shortformat(self, config, shortformat):
        if shortformat:
            return {
                'widget': shortformat.widget,
                'shorthelp': unicode(shortformat.shorthelp(config))
            }
        else:
            return None

    def number_of_groups_where_is_examiner(self, instance):
        if isinstance(instance, self.model):
            return Examiner.objects.filter(
                user = self.view.request.user,
                assignmentgroup__parentnode=instance).count()

    def gradingsystemplugin_title(self, instance):
        if isinstance(instance, self.model):
            try:
                pluginapi = instance.get_gradingsystem_plugin_api()
            except GradingSystemPluginNotInRegistryError:
                return None
            else:
                return unicode(pluginapi.title)



class ListOrCreateAssignmentRest(BaseNodeListOrCreateView):
    """
    List the subjects where the authenticated user is admin.
    """
    permissions = (IsAuthenticated,)
    resource = AssignmentResource

    def authenticate_postrequest(self, user, parentnode_id):
        periodadmin_required(user, parentnode_id)

    def get_queryset(self):
        qry = super(ListOrCreateAssignmentRest, self).get_queryset()
        qry = qry.order_by('-publishing_time')
        return qry


class InstanceAssignmentRest(BaseNodeInstanceModelView):
    """
    This API provides read, update and delete on a single subject.
    """
    permissions = (IsAuthenticated, IsAssignmentAdmin)
    resource = AssignmentInstanceResource

    def get_queryset(self):
        qry = super(InstanceAssignmentRest, self).get_queryset()
        qry = qry.select_related('parentnode', 'parentnode__parentnode')
        qry = qry.prefetch_related('admins', 'admins__devilryuserprofile',
            'parentnode__admins', 'parentnode__admins__devilryuserprofile',
            'parentnode__parentnode__admins', 'parentnode__parentnode__admins__devilryuserprofile')
        qry = qry.annotate(number_of_groups=Count('assignmentgroups', distinct=True))
        qry = qry.annotate(number_of_deliveries=Count('assignmentgroups__deadlines__deliveries', distinct=True))
        qry = qry.annotate(number_of_candidates=Count('assignmentgroups__candidates', distinct=True))
        return qry
